pkg/gardenlet/controller/bastion/bastion_reconciler.go

```diff
@@ -131,7 +131,7 @@ func (r *reconciler) reconcileBastion(
 
 	// ensure finalizer is set
 	if !controllerutil.ContainsFinalizer(bastion, finalizerName) {
-		if err := controllerutils.PatchAddFinalizers(ctx, gardenClient, bastion, finalizerName); err != nil {
+		if err := controllerutils.StrategicMergePatchAddFinalizers(ctx, gardenClient, bastion, finalizerName); err != nil {
```

Comment:
I think the following lines below are very dubious:
```
		// the patch above already triggers a reconcile, so we can stop here
		return nil
```
I think none of the other controllers has similar behavior. Might it be that it actually relies on the non-merge patch? Could you perhaps double-check this, or simply remove this `return nil` it looks somehow wrong anyway?

Reply:
Nice catch, thank you!
TBH, I didn't pay much attention to this comment, but you're right. It seems like this code is relying on either an update watch event caused by the finalizer or a conflict from `PatchAddFinalizer`. But now, there will be no more conflict.
I will remove the `return nil` to make the behavior similar to all other controllers.

Reply:
Ok, one step back. While addressing your general question, I noticed that the BackupEntry and Shoot controller are actually doing the same thing: https://github.com/gardener/gardener/blob/38ad556a7dec689b35c3ee6dd3cf59bd64b5c705/pkg/gardenlet/controller/shoot/shoot_control.go#L429-L434

And it seems to be for a good reason – to prevent unwanted double reconciliations, see https://github.com/gardener/gardener/pull/2364

Considering this, I think we should actually not merge `ContainsFinalizer` and `StrategicMergePatchAddFinalizers` to a single func, because we should actually distinguish between wether the finalizer was added or was already present.
And we should probably return after adding the finalizer (like it is done here and in the other 2 controlelrs) in all controllers.
WDYT?

Reply:
Alternatively, we could think about changing the "predicate" for enqueuing on update events:
instead of enqueuing when the generation is not equal to the observed generation, we could enqueue if the generation changed.
BTW: this is also what controller-runtime supports out of the box with `predicate.GenerationChangedPredicate`.

Reply:
Maybe @ialidzhikov has some additional thoughts on this as author of https://github.com/gardener/gardener/pull/2364

Reply:
> And it seems to be for a good reason – to prevent unwanted double reconciliations, see #2364

Yes, as you stated the reason is to prevent any unwanted additional reconciliations. As described in https://github.com/gardener/gardener/pull/2364, on Shoot creation we have update/patch requests (like the one that adds a finalizer) that occur when metadata.generation != status.observedGeneration which is causing the obj to be requeued again. So currently on the Shoot reconciliation we use such "early exit" approach (`return nil` ) to prevent additional reconciliation (can be translated also like "abort the current reconciliation and rely on the next one that is caused by the updated that was just executed)".

You can check the description of https://github.com/gardener/gardener/pull/2364 for even better explanation.
So bottom line, we should make sure that we don't have (or introduce) additional reconciliations because of updates on existing creation/reconciliation.

PS. Since some time, in my local setup I notice again that on Shoot creation the Shoot gets reconciled again after 
A successful creation. I added this to my list, but still didn't manage to have a look.

Reply:
It's a bit confusing for me that only a few controllers do this, and the others not. I would try to make it consistent across controllers, if it's doesn't hurt too much. Does this even work properly with `StrategicMergePatchAddFinalizers` instead of `PatchAddFinalizers`?

> Alternatively, we could think about changing the "predicate" for enqueuing on update events:
instead of enqueuing when the generation is not equal to the observed generation, we could enqueue if the generation changed.

I like this idea, is it too complex a change? If yes, perhaps we could merge this PR as it is and do this refactoring in a separate PR.

Reply:
Thanks for your input!

> Does this even work properly with `StrategicMergePatchAddFinalizers` instead of `PatchAddFinalizers`?

I think, this should work as well with strategic merge patches. The only difference is, that we now don't get a conflict error anymore, if we read a stale object from the cache without a finalizer and then try to add it again. In that case we will simply return `nil` instead of a conflict error. So it will not go into exponential backoff but wait for the watch event instead before retrying. But this is also what happens currently with merge patch, when the first try to add the finalizer is successful and we return `nil` afterwards.

> So bottom line, we should make sure that we don't have (or introduce) additional reconciliations because of updates on existing creation/reconciliation.

> I like this idea, is it too complex a change? If yes, perhaps we could merge this PR as it is and do this refactoring in a separate PR.

I would like to do the following: we proceed with this PR as is (+ the commit for addressing the logging suggestions, will push it next), but we keep the current logic wrt to checking the finalizers first with `ContainsFinalizer` and then using `StrategicMergePatchAddFinalizers`. 
I will also keep the "special logic" in the shoot, bastion, etc. controller and not touch it there or in any other controllers to not regress anything in this PR.

And then, we can carefully think about the best way to prevent duplicated reconciliations in all controllers and check the duplicated shoot reconciliations that @ialidzhikov mentioned again.

Let me know, if you have any objections.

Reply:
No objections, thanks!